### ISSUE TITLE
Enquiring about the proper way for github-push-action to actually fire a workflow (in an organization environment)

### ISSUE THREAD

**Issue by rturrado** [ad-m/github-push-action#172]
Hi all,

I am using the following code at the moment in one of my workflows:

```
       - name: Commit files
        run: |
          git config --local user.email "41898282+github-actions[bot]@users.noreply.github.com"
          git config --local user.name "github-actions[bot]"
          git commit -a -m "Updating v3_distro with new generated files from v3"
      - name: Push changes
        uses: ad-m/github-push-action@master
        with:
          force_with_lease: true
```

This is correctly pushing some files to a given branch.

The current workflow that I have is: on a given condition on branch `A`, check out branch `B`, update some files, commit and push.

However, that is not firing a second workflow on branch `B`, which I would like to do a build and test. That second workflow firing section looks like this:

```
on:
  push:
    branches:
    - B
```

1) I understand this is happening because I'm pushing without any type of authentication. Is that correct?
2) If that's so: let's suppose I want to add a token/PAT authentication, is this a per user thing (e.g. all users pushing to `A` will need to have a token/PAT), or can it be set at an organization level?
3) Finally: for what I'm trying to do, firing this second workflow on a different branch, will I need to use both the token and the PAT?

Many thanks,
Roberto.

**Comment by ZPascal** [comment]
Hi @rturrado,

>1. I understand this is happening because I'm pushing without any type of authentication. Is that correct?

You're using the default GITHUB_TOKEN? Then you are not [triggering](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#using-the-github_token-in-a-workflow) the workflow inside branch B at the end.

>2. If that's so: let's suppose I want to add a token/PAT authentication, is this a per user thing (e.g. all users pushing to A will need to have a token/PAT), or can it be set at an organization level?

It is not necessary that all users use PATs, and it is still possible to handle pushes to Branch A from the Allowed Users group in the way you prefer, e.g. via SSH. Integration at the organizational level and using the PAT for workflow runs are also possible. I recommend using a technical user for the process in general. 

**FYI:** If you want to push to a protected branch, the use of the owner's PAT is required.

> 3. Finally: for what I'm trying to do, firing this second workflow on a different branch, will I need to use both the token and the PAT?

It is possible to use both tokens, e.g. GITHUB_TOKEN for checkout and PAT for push.

Regards,
Pascal

**Comment by rturrado** [comment]
Hi @ZPascal,

Many thanks for the quick reply. Commenting below.

> You're using the default GITHUB_TOKEN? Then you are not [triggering](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#using-the-github_token-in-a-workflow) the workflow inside branch B at the end.

The `default GITHUB_TOKEN` is the `${{ secrets.GITHUB_TOKEN }}` that any workflow defines by default, right? If you see the `Push changes` step in my original post, I am just doing a push with a `force_with_lease`.

> It is not necessary that all users use PATs, and it is still possible to handle pushes to Branch A from the Allowed Users group in the way you prefer, e.g. via SSH. Integration at the organizational level and using the PAT for workflow runs are also possible. I recommend using a technical user for the process in general.

What's the `Allowed Users` group? I cannot understand much of what you say, sorry. I don't know if you are understanding what I'm trying to do. This is what I have:

- A `company/repo` GitHub repo.
- A couple of branches `A` and `B` in the repo.
- A workflow that is fired with pushes to branch `A` and ends up using this `github-push-action` to push into branch `B`.
- A workflow that is fired with pushes to branch `B`.

I have tried `github-push-action` with 1) no tokens at all and `force_with_lease: true`, and 2) with `github_token: ${{ secrets.GITHUB_TOKEN }}`, and none of this ways make the workflow on branch `B` to be fired. However, the pushes are done.

> **FYI:** If you want to push to a protected branch, the use of the owner's PAT is required.

OK.

> > 3. Finally: for what I'm trying to do, firing this second workflow on a different branch, will I need to use both the token and the PAT?
> 
> It is possible to use both tokens, e.g. GITHUB_TOKEN for checkout and PAT for push.

Why would I need `GITHUB_TOKEN` for the checkout?

Best regards,
Roberto.

**Comment by ZPascal** [comment]
Hi @rturrado,

>The default GITHUB_TOKEN is the ${{ secrets.GITHUB_TOKEN }} that any workflow defines by default, right? If you see the Push changes step in my original post, I am just doing a push with a force_with_lease.

Yes, that's correct. We're using inside the `github-push-action` the default `GITHUB_TOKEN` also known as `${{ secrets.GITHUB_TOKEN }}`. You can find more information about the `GITHUB_TOKEN` [here](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#about-the-github_token-secret).

>What's the Allowed Users group?

The group of users who got access to the repo.

>I don't know if you are understanding what I'm trying to do.

I'm understanding your case.

>I have tried github-push-action with
no tokens at all and force_with_lease: true, and
with github_token: ${{ secrets.GITHUB_TOKEN }},
and none of these ways make the workflow on branch B to be fired. However, the pushes are done.

Yes, that makes sense, because you are using at the end the same `GITHUB_TOKEN`. This token can not trigger the workflow!

>Why would I need GITHUB_TOKEN for the checkout?

You don't need it. You can handle all GIT interactions with the same token e.g. PAT.

To summarize your situation. it's necessary that you create a PAT to handle the push.

Regards,
Pascal

**Comment by ZPascal** [comment]
@rturrado Do you have any further questions, or can we close this case?

**Comment by rturrado** [comment]
Hi @ZPascal,

Actually, I have another question.

I have the following setup which is working fine (i.e. it's doing the `git push`, although I still have to use the token to fire a second workflow):
1. Workflow `W1` is fired on branch `A`, checks out branch `A`, and uploads some artifacts.
2. Workflow `W2` is fired on branch `A`, waits for `W1` to finish, checks out branch `B`, then downloads the artifacts, copies them to the workspace, and does the `git push` to branch `B`.

Now, I've tried to modify it to use only one workflow with two jobs, but then the `git push` is telling me there is nothing to push:
1. Workflow `W` is fired on branch `A`, and it contains two jobs, `J1`, and `J2`.
2. Job `J1` checks out branch `A`, and uploads some artifacts.
3. Job `J2` waits for `W1` to finish, checks out branch `B`, then downloads the artifacts, copies them to the workspace, and does the `git push` to branch `B`.

Do you reckon this may have something to do with the `github-push-action`?

Thanks,
Roberto.

**Comment by rturrado** [comment]
> Hi @rturrado, 
> > Do you reckon this may have something to do with the github-push-action?
> 
> I don't think so, but I try to support you. In general, I recommend opening a new issue and close this case. Please add also the push output of the action and the git commit and push workflow part to the new issue.
> 
> Regards,
> Pascal 
> 

Many thanks @ZPascal, we can close this issue then.
